kernel/src/scan/mod.rs

```diff
@@ -218,20 +214,34 @@ impl Scan {
                 let read_result = if partition_fields.is_empty() {
                     read_result
                 } else {
-                    let mut fields = Vec::with_capacity(partition_fields.len() + len);
+                    debug!("Adding back in partition columns");
+                    let final_schema = self.schema();
+                    let mut fields = vec![None; final_schema.fields.len()];
+                    for select_field in read_schema.fields() {
+                        let field_position = final_schema
+                            .field_index(&select_field.name)
+                            .ok_or_else(|| Error::missing_column(&select_field.name))?;
+                        fields[field_position] = Some(Expression::column(select_field.name()));
+                    }
                     for field in &partition_fields {
                         let value_expression = parse_partition_value(
                             add.partition_values.get(field.name()),
                             field.data_type(),
                         )?;
-                        fields.push(Expression::Literal(value_expression));
+                        let field_position = final_schema
+                            .field_index(&field.name)
+                            .ok_or_else(|| Error::missing_column(&field.name))?;
+                        fields[field_position] = Some(Expression::Literal(value_expression));
                     }
-                    fields.extend(select_fields.clone());
-
+                    debug!("Final fields to evaluate for partition columns: {fields:?}");
+                    let fields = fields
+                        .into_iter()
+                        .map(|opt| opt.ok_or_else(|| Error::generic("Invalid columns in select")))
```

Comment:
This error isn't helpful (no column name), so we may as well just verify `len()` after flattening the list?
```rust
let fields = fields.into_iter().flatten().collect();
if fields.len() != final_schema.len() {
  // the unthinkable happened!
}
```

Reply:
not doing this anymore, so resolving.